lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -3312,18 +3341,30 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     // `originalResults` and `seed`.
     auto adjParamArgs = getAdjoint().getArgumentsWithoutIndirectResults();
     auto origNumParams = origConv.getNumParameters();
+    auto origNumParamsWithoutSelf = origTy->hasSelfParam() ? origNumParams - 1
+                                                           : origNumParams;
     auto origNumResults = origTy->getNumResults();
     // The adjoint function has type
-    //   (arg0, ..., argn, pv0, ..., pvn, origres, seed) -> (arg0, ..., argn).
+    //   (arg0, ..., argn, pv0, ..., pvn, origres, seed, [self])
+    //       -> ([self], [arg0], ..., [argn]).
+    // Square brackets denote [] elements that are not always in the signature:
+    //   * "self" is present in the argument list when it's present in the
+    //      original's argument list.
+    //   * Results are present when the adjoint is with respect to the
+    //     corresponding original argument.
+    //
     // We get each range of arguments by shifting the `paramArgsData` pointer.
     auto *paramArgsData = adjParamArgs.data();
-    originalParametersInAdj = {paramArgsData, origNumParams};
-    paramArgsData += origNumParams;
+    originalParametersInAdj.reserve(origNumParams);
+    for (unsigned i = 0; i < origNumParamsWithoutSelf; i++)
```

Comment:
Adjust this to the local convention in this file:
```cpp
for (auto i : range(origNumParamsWithoutSelf))
```

AD's code base tends to use LLVM range helpers whenever possible.